src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -203,10 +206,13 @@ public String url() {
      *  Note: to avoid deadlocks, when nesting locks we ALWAYS need to lock on the guard first, THEN the WorkflowRun.
      *  Synchronizing this helps ensure that fields are not mutated during a {@link #save()} operation, since that locks on the Run.
```

Comment:
I don't totally understand this line. I would like to delete `metadataGuard` completely and just synchronize on `this` instead when modifying metadata, so that we don't have to worry about potential deadlock, but I am concerned that it would cause performance issues.

Reply:
> I would like to delete `metadataGuard` completely and just synchronize on `this` instead when modifying metadata, so that we don't have to worry about deadlock

I am tempted to do the same. I would hope that no one is holding a lock on `WorkflowRun` for very long, but not sure.